### PR TITLE
src/arch/simddetect.cpp: fix NEON detection on FreeBSD

### PR DESCRIPTION
**Pull request by clausecker**
FreeBSD does not have the `<asm/hwcap>` header, requiring some slightly different code to check for NEON availability.  This code might also apply to other BSD-like operating systems such as NetBSD, OpenBSD, DragonflyBSD, or iOS, but I haven't checked.

### CODE REVIEW

File: src/arch/simddetect.cpp [hunk 1]
```diff
@@ -85,7 +87,7 @@ SIMDDetect SIMDDetect::detector;
 bool SIMDDetect::neon_available_ = true;
 #elif defined(HAVE_NEON)
 // If true, then Neon has been detected.
-bool SIMDDetect::neon_available_;
+bool SIMDDetect::neon_available_ = false;
```
Comment (stweil) [inline src/arch/simddetect.cpp:90]: Why this change?

File: src/arch/simddetect.cpp [hunk 2]
```diff
@@ -55,10 +55,12 @@
 #if defined(HAVE_NEON) && !defined(__aarch64__)
 #  ifdef ANDROID
 #    include <cpu-features.h>
-#  else
-/* Assume linux */
+#  elif defined(__linux__)
 #    include <asm/hwcap.h>
 #    include <sys/auxv.h>
+#  elif defined(__FreeBSD__)
```
Comment (stweil) [inline src/arch/simddetect.cpp:61]: I'd prefer to check for Android first, then FreeBSD, then the rest (no explicit check for Linux). That's closer to the current code and perhaps covers more cases.

File: src/arch/simddetect.cpp [hunk 3]
```diff
@@ -216,9 +218,12 @@ SIMDDetect::SIMDDetect() {
     if (family == ANDROID_CPU_FAMILY_ARM)
       neon_available_ = (android_getCpuFeatures() & ANDROID_CPU_ARM_FEATURE_NEON);
   }
-#  else
-  /* Assume linux */
+#  elif defined(__linux__)
```
Comment (stweil) [inline src/arch/simddetect.cpp:221]: See comment above.

**Review by stweil — CHANGES_REQUESTED** [review]

File: src/arch/simddetect.cpp (hunk 1 above)
Comment (clausecker) [reply src/arch/simddetect.cpp:90]: This change makes it so that when compiling for an unknown operating system, `neon_available_` defaults to `false` instead of the build failing.  It is connected to the changed further below where now `neon_available_` is no longer set when compiling for an unknown operating system.

File: src/arch/simddetect.cpp (hunk 2 above)
Comment (clausecker) [reply src/arch/simddetect.cpp:61]: If you do it that way, the build will fail on non-Linux operating systems.  I have specifically designed this patch to permit compilation on operating systems for which we don't know how to obtain the ELF aux vectors, falling back to the safe assumption that NEON is not supported.

File: src/arch/simddetect.cpp (hunk 1 above)
Comment (stweil) [reply src/arch/simddetect.cpp:90]: Without your change it is also initialized to `false`.

Comment (clausecker) [reply src/arch/simddetect.cpp:90]: Yes.  This initialisation is overridden in the constructor if running on a platform on which we know how to detect the availability of NEON.

File: src/arch/simddetect.cpp (hunk 2 above)
Comment (stweil) [reply src/arch/simddetect.cpp:61]: I don't think that the code should silently assume that NEON is unsupported because that might result in code which is slower than necessary. Failing seems better to me, because it allows investigating and fixing the problem.

Comment (stweil) [reply src/arch/simddetect.cpp:61]: As we are talking about 32 bit ARM systems (which typically are not very fast), using NEON is especially important to get at least an acceptable performance.

Comment (clausecker) [reply src/arch/simddetect.cpp:61]: Ideally, a configure test would be used to check which way of finding NEON support is available.  Just failing the build seems nasty to users of other operating systems.  I mean I can change the patch like that if you insist.

Comment (stweil) [reply src/arch/simddetect.cpp:61]: Yes, a configure test (which only runs if necessary) would be the best solution.

Comment (clausecker) [reply src/arch/simddetect.cpp:61]: I've never designed a configure test before.  I'll try to figure it out and report back.

Comment (clausecker) [reply src/arch/simddetect.cpp]: I have added configure tests as directed.  Please also check and revise the `AC_CONDITIONAL` usage.  I don't think these are meant to be redefined.

File: src/arch/simddetect.cpp (hunk 1 above)
Comment (stweil) [inline src/arch/simddetect.cpp:90]:
```suggestion
bool SIMDDetect::neon_available_;
```

File: configure.ac [hunk 4]
```diff
@@ -163,7 +163,7 @@ case "${host_cpu}" in
 
     ;;
 
-  aarch64)
+  aarch64|arm64)
```
Comment (stweil) [inline configure.ac:166]: Which platforms use `arm64` instead of `aarch64`?

File: configure.ac [hunk 5]
```diff
@@ -178,6 +178,7 @@ case "${host_cpu}" in
       AC_DEFINE([HAVE_NEON], [1], [Enable NEON instructions])
       NEON_CXXFLAGS="-mfpu=neon"
       AC_SUBST([NEON_CXXFLAGS])
+      check_for_neon=1
```
Comment (stweil) [inline configure.ac:181]: I suggest to include the code from below here. Then there is no need for `check_for_neon`.

File: src/arch/simddetect.cpp [hunk 6]
```diff
@@ -210,15 +212,18 @@ SIMDDetect::SIMDDetect() {
 #endif
 
 #if defined(HAVE_NEON) && !defined(__aarch64__)
-#  ifdef ANDROID
+#  if HAVE_ANDROID_GETCPUFAMILY
```
Comment (stweil) [inline src/arch/simddetect.cpp:215]:
```suggestion
#  if defined(HAVE_ANDROID_GETCPUFAMILY)
```

Comment (stweil) [reply src/arch/simddetect.cpp:215]: `HAVE_ANDROID_GETCPUFAMILY` is not always defined, so a simple `#if HAVE_ANDROID_GETCPUFAMILY` could trigger compiler warnings.

File: src/arch/simddetect.cpp [hunk 7]
```diff
@@ -210,15 +212,18 @@ SIMDDetect::SIMDDetect() {
...
   {
     AndroidCpuFamily family = android_getCpuFamily();
     if (family == ANDROID_CPU_FAMILY_ARM)
       neon_available_ = (android_getCpuFeatures() & ANDROID_CPU_ARM_FEATURE_NEON);
   }
-#  else
-  /* Assume linux */
+#  elif HAVE_GETAUXVAL
```
Comment (stweil) [inline src/arch/simddetect.cpp:221]:
```suggestion
#  elif defined(HAVE_GETAUXVAL)
```

File: src/arch/simddetect.cpp [hunk 8]
```diff
@@ -210,15 +212,18 @@ SIMDDetect::SIMDDetect() {
...
   neon_available_ = getauxval(AT_HWCAP) & HWCAP_NEON;
+#  elif HAVE_ELF_AUX_INFO
```
Comment (stweil) [inline src/arch/simddetect.cpp:223]:
```suggestion
#  elif defined(HAVE_ELF_AUX_INFO)
```

File: configure.ac (hunk 4 above)
Comment (clausecker) [reply configure.ac:166]: FreeBSD does for example.  I have not checked other platforms.

File: src/arch/simddetect.cpp (hunk 6 above)
Comment (clausecker) [reply src/arch/simddetect.cpp:215]: An undefined macro evaluates to `0` in a `#if` preprocessor test.  No warning is emitted in this case.

File: configure.ac (hunk 5 above)
Comment (clausecker) [reply configure.ac:181]: Which code do you mean specifically?

**Comment by clausecker** [comment]
Please consider opening a code review next time so the system doesn't send me one email for each comment you wrote.

File: src/arch/simddetect.cpp (hunk 6 above)
Comment (stweil) [reply src/arch/simddetect.cpp:215]: The GNU C++ compiler and clang emit warnings (depending on warning level). See https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html (-Wundef).

File: configure.ac (hunk 5 above)
Comment (stweil) [reply configure.ac:181]: Lines 197–202.

Comment (clausecker) [reply configure.ac:181]: Line 196 checks for `check_for_neon`.  The point of this variable is to only run the NEON checks on platforms where NEON could possibly be available.  You [specifically requested](https://github.com/tesseract-ocr/tesseract/pull/3782#discussion_r841081250) that the test be implemented that way.

Comment (stweil) [reply configure.ac:181]: Right. And now I suggest to move lines 197–202 to replace line 181 in your commit. That also runs the NEON checks only when needed, but simplifies and reduces the code for that.

Comment (clausecker) [reply configure.ac:181]: There was some sort of reason why I did it the way I did it.  I can try to check if it works the way you indicate, but it has been almost two months since I last touched this changeset and it will be difficult for me to re-run all the tests and ensure it still works correctly.  It would be a lot easier for me just to commit it as is.

Comment (stweil) [reply configure.ac:181]: Then please add at least all other suggested changes.

File: src/arch/simddetect.cpp [hunk 9]
```diff
@@ -53,12 +53,14 @@
 #endif
 
 #if defined(HAVE_NEON) && !defined(__aarch64__)
-#  ifdef ANDROID
+#  if HAVE_ANDROID_GETCPUFAMILY
```
Comment (stweil) [inline src/arch/simddetect.cpp:56]:
```suggestion
#  if defined(HAVE_ANDROID_GETCPUFAMILY)
```

File: src/arch/simddetect.cpp [hunk 10]
```diff
@@ -53,12 +53,14 @@
...
 #    include <cpu-features.h>
-#  else
-/* Assume linux */
+#  elif HAVE_GETAUXVAL
```
Comment (stweil) [inline src/arch/simddetect.cpp:58]:
```suggestion
#  elif defined(HAVE_GETAUXVAL)
```

File: src/arch/simddetect.cpp [hunk 11]
```diff
@@ -53,12 +53,14 @@
...
 #    include <asm/hwcap.h>
 #    include <sys/auxv.h>
+#  elif HAVE_ELF_AUX_INFO
```
Comment (stweil) [inline src/arch/simddetect.cpp:61]:
```suggestion
#  elif defined(HAVE_ELF_AUX_INFO)
```

**Review by stweil — APPROVED** [review]
Thanks!

**Comment by stweil** [comment]
While this pull request was an important fix for ARM hosts, we had a similar severe problem for amd64 / x86_64 host. See issue #3964.
